go/chat/search/storage.go

```diff
@@ -48,204 +70,350 @@ func newStore(g *globals.Context) *store {
+	if err != nil {
+		return nil, err
 	}
-	for _, t := range entry.Alias {
-		ret.Alias[t.Alias] = make(map[string]chat1.EmptyStruct, len(t.Tokens))
-		for _, token := range t.Tokens {
-			ret.Alias[t.Alias][token] = chat1.EmptyStruct{}
+	if !found {
+		return newAliasEntry(), nil
+	}
+	res = &ae
+	if res.Version != refAliasEntry.Version {
+		return newAliasEntry(), nil
+	}
+	return res, nil
+}
+
+func (s *store) putTokenEntry(ctx context.Context, uid gregor1.UID, convID chat1.ConversationID,
+	token string, te *tokenEntry) (err error) {
+	defer func() {
+		if err == nil {
+			s.tokenCache.Add(s.tokenCacheKey(uid, convID, token), te)
 		}
+	}()
+	key, err := s.tokenKey(ctx, uid, convID, token)
+	if err != nil {
+		return err
 	}
-	for _, msgID := range entry.Metadata.SeenIDs {
-		ret.Metadata.SeenIDs[msgID] = chat1.EmptyStruct{}
+	return s.edb.Put(ctx, key, te)
+}
+
+func (s *store) putAliasEntry(ctx context.Context, alias string, ae *aliasEntry) error {
+	key, err := s.aliasKey(ctx, alias)
+	if err != nil {
+		return err
 	}
+	s.aliasCache.Remove(key)
+	return s.edb.Put(ctx, key, ae)
+}
 
-	return ret, nil
+func (s *store) deleteTokenEntry(ctx context.Context, uid gregor1.UID, convID chat1.ConversationID,
+	token string) {
+	key, err := s.tokenKey(ctx, uid, convID, token)
+	if err != nil {
+		s.Debug(ctx, "deleteTokenEntry: failed to get token key: %s", err)
+		return
+	}
+	s.tokenCache.Remove(s.tokenCacheKey(uid, convID, token))
+	if err := s.G().LocalChatDb.Delete(key); err != nil {
+		s.Debug(ctx, "deleteTokenEntry: failed to delete key: %s", err)
+	}
 }
 
-func (s *store) putLocked(ctx context.Context, convID chat1.ConversationID, uid gregor1.UID, idx *chat1.ConversationIndex) error {
-	if idx == nil {
-		return nil
-	}
-	dbKey := s.dbKey(convID, uid)
-	entry := chat1.ConversationIndexDisk{
-		Index: make([]chat1.TokenTuple, len(idx.Index)),
-		Alias: make([]chat1.AliasTuple, len(idx.Alias)),
-		Metadata: chat1.ConversationIndexMetadataDisk{
-			Version: IndexVersion,
-			SeenIDs: make([]chat1.MessageID, len(idx.Metadata.SeenIDs)),
-		},
-	}
-
-	i := 0
-	for token, msgIDMap := range idx.Index {
-		msgIDs := make([]chat1.MessageID, len(msgIDMap))
-		ii := 0
-		for msgID := range msgIDMap {
-			msgIDs[ii] = msgID
-			ii++
-		}
-		entry.Index[i] = chat1.TokenTuple{Token: token, MsgIDs: msgIDs}
-		i++
+func (s *store) deleteAliasEntry(ctx context.Context, alias string) {
+	key, err := s.aliasKey(ctx, alias)
+	if err != nil {
+		s.Debug(ctx, "deleteAliasEntry: failed to get key: %s", err)
+		return
 	}
+	s.aliasCache.Remove(key)
+	if err := s.G().LocalChatDb.Delete(key); err != nil {
+		s.Debug(ctx, "deleteAliasEntry: failed to delete key: %s", err)
+	}
+}
 
-	j := 0
-	for alias, tokenMap := range idx.Alias {
-		tokens := make([]string, len(tokenMap))
-		jj := 0
-		for token := range tokenMap {
-			tokens[jj] = token
-			jj++
-		}
-		entry.Alias[j] = chat1.AliasTuple{Alias: alias, Tokens: tokens}
-		j++
+type addTokenBatch struct {
```

Comment:
why not have the `Put` methods just write to memory and then have periodic flushing to disk? would be a bigger win when reindexing a bunch of pages of messages

Reply:
Don't want any chance that MD gets out of sync with these keys.